src/server/shell/basic_idle_handler.cpp

```diff
@@ -0,0 +1,181 @@
+    {
+        return buffer_;
+    }
+
+    auto screen_position() const -> geom::Rectangle override
+    {
+        return {{-coverage_size / 2, -coverage_size / 2}, {coverage_size, coverage_size}};
+    }
+
+    auto clip_area() const -> std::experimental::optional<geom::Rectangle> override
+    {
+        return {};
+    }
+
+    auto alpha() const -> float override
+    {
+        return 0.5f;
+    }
+
+    auto transformation() const -> glm::mat4 override
+    {
+        return glm::mat4{1};
+    }
+
+    auto shaped() const -> bool override
+    {
+        return false;
+    }
+
+private:
+    std::shared_ptr<mg::Buffer> const buffer_;
+};
+
+struct Dimmer : ms::IdleStateObserver
+{
+    Dimmer(
+        std::shared_ptr<mi::Scene> const& input_scene,
+        std::shared_ptr<mg::GraphicBufferAllocator> const& allocator)
+        : input_scene{input_scene},
+          allocator{allocator}
+    {
+    }
+
+    void active() override
+    {
+        std::lock_guard<std::mutex> lock{mutex};
+        if (renderable)
+        {
+            input_scene->remove_input_visualization(renderable);
+            renderable.reset();
+        }
+
+    }
+
+    void idle() override
+    {
+        std::lock_guard<std::mutex> lock{mutex};
+        if (!renderable)
+        {
+            renderable = std::make_shared<DimmingRenderable>(*allocator);
+            input_scene->add_input_visualization(renderable);
+        }
+    }
+
+private:
+    std::shared_ptr<mi::Scene> const input_scene;
+    std::shared_ptr<mg::GraphicBufferAllocator> const allocator;
+    std::mutex mutex;
+    std::shared_ptr<mg::Renderable> renderable;
+};
+
+struct PowerModeSetter : ms::IdleStateObserver
+{
+    PowerModeSetter(std::shared_ptr<msh::DisplayConfigurationController> const& controller, MirPowerMode power_mode)
+        : controller{controller},
+          power_mode{power_mode}
+    {
+    }
+
+    void active() override
+    {
+        controller->set_power_mode(mir_power_mode_on);
+    }
+
+    void idle() override
+    {
+        controller->set_power_mode(power_mode);
+    }
+
+private:
+    std::shared_ptr<msh::DisplayConfigurationController> const controller;
+    MirPowerMode const power_mode;
+};
+}
+
+msh::BasicIdleHandler::BasicIdleHandler(
+    std::shared_ptr<ms::IdleHub> const& idle_hub,
+    std::shared_ptr<input::Scene> const& input_scene,
+    std::shared_ptr<graphics::GraphicBufferAllocator> const& allocator,
+    std::shared_ptr<msh::DisplayConfigurationController> const& display_config_controller)
+    : idle_hub{idle_hub},
+      timeouts{
+          Timeout{5s, std::make_shared<Dimmer>(input_scene, allocator)},
+          Timeout{10s, std::make_shared<PowerModeSetter>(display_config_controller, mir_power_mode_standby)},
+          Timeout{15s, std::make_shared<PowerModeSetter>(display_config_controller, mir_power_mode_suspend)},
+          Timeout{20s, std::make_shared<PowerModeSetter>(display_config_controller, mir_power_mode_off)},
```

Comment:
These seem like pretty aggressive defaults? Dim the screen after 5 seconds of inactivity, turn it off after 20? That's much more aggressive than my phone :)

Reply:
Ah yes, that was just for testing purposes. This really should be configurable.